mono/mini/method-to-ir.c

```diff
@@ -10055,6 +10055,21 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 					MonoClass *tclass = mono_class_from_mono_type_internal ((MonoType *)handle);
 
 					mono_class_init_internal (tclass);
+
+					// Optimize to true/false if next instruction is `call instance bool Type::get_IsValueType()`
+					guchar *is_vt_ip;
+					guint32 is_vt_token;
+					if ((is_vt_ip = il_read_call (next_ip + 5, end, &is_vt_token)) && ip_in_bb (cfg, cfg->cbb, is_vt_ip)) {
+						MonoMethod *is_vt_method = mini_get_method (cfg, method, is_vt_token, NULL, generic_context);
+						if (is_vt_method->klass == mono_defaults.systemtype_class && !strcmp ("get_IsValueType", is_vt_method->name)) {
+							next_ip = is_vt_ip;
+							EMIT_NEW_ICONST (cfg, ins, m_class_is_valuetype (tclass) ? 1 : 0);
```

Comment:
Yea this seems wrong for gsharedvt: `tclass` will be some generic type param; also it could be instantiated with either a class or a valuetype.  You'd need some kind of runtime check in that case.

Reply:
Ok, the optimization now gives up on such types (and uses the default "slow" path)